grudge/dt_utils.py

```diff
@@ -45,22 +57,58 @@
 from pytools import memoize_on_first_arg
 
 
+def estimate_local_timestep(
+        dcoll: DiscretizationCollection, wavespeed) -> DOFArray:
+    r"""Computes an estimate for the local timestep size at each node.
+    The estimate is obtained using the following formula:
+
+    .. math::
+
+        \Delta t_{loc} = \operatorname{min}\left(\Delta r_i\right)\frac{r_D}{|c|}
+
+    where :math:`\operatorname{min}\left(\Delta r_i\right)` is the minimum
+    node distance on the reference cell (see :func:`dt_non_geometric_factors`),
+    :math:`r_D` is the inradius of the physical cell
+    (see :func:`dt_geometric_factors`), and :math:`|c|` is the local
+    magnitude of the characteristic wavespeed.
+
+    :arg wavespeed: a number or :class:`~meshmode.dof_array.DOFArray` for the
+        characteristic wavespeed.
```

Comment:
I wonder if we should make the wavespeed optional, to allow use of the returned quantity as a replacement of `h/N^2`, e.g. in the formula given by @MTCam in this comment: https://github.com/illinois-ceesd/mirgecom/pull/368#discussion_r641814288

Reply:
That would be great.  In general, and imo much to the credit of code reviews in our current effort, I've been trying to avoid making functions take arguments that are used for scaling only:

e.g., I find:
`a = fac * f(x, y, z)` 

... is superior to:
`a = f(x, y, z, fac)`
(with the scaling on the inside)

The former is easier to read and understand immediately what is happening.  The current function is no exception.  The physics user developers will understand as much as they need from:

`dt = grudge.dt_facs/wavespeed`, while something like: `grudge.estimate_local_timestep(discr, wavespeed)` is a bit more opaque, requiring a bit more digging to figure out how the timestep is estimated.   

This clarity advantage is in addition to the thing that @inducer mentions above.  I would like to write something like:

`dt = grudge.dt_facs/(wavespeed + mu/grudge.dt_facs)` 

... for a quick and dirty viscous timestep estimate.

Reply:
So I've added @inducer's suggestion above https://github.com/inducer/grudge/pull/121/commits/cd2740380c16884db521ee86fc1e98c9608eabc8. But do you both (@MTCam + @inducer) think this should be renamed to `dt_factors` or something like that and always return just the dt scaling factor? I'm trying to get a sense on what you both prefer. Currently, to do what @MTCam wants, you'd basically write:
```python
dt_factors = estimate_local_timestep(dcoll)

dt = dt_factors/(wavespeed + mu/dt_factors)
```

Reply:
Thanks a bunch, @thomasgibson! I defer to your judgements about the naming of it.  One thing you *could* do is provide a public function something like: `get_characteristic_length_scale` and then your `estimate_local_timestep` can use that, or users can directly use the `get_characteristic_length` or whatever if they want.

Reply:
I like `characteristic_length`. I'll leave it up to you whether to keep `estimate_local_timestep` as a wrapper, but my gut feeling is that
```
dt = characteristic_length(...) / wavespeed
```
is about the same length and perhaps clearer than `estimate_local_timestep(..., wavespeed)`.

Reply:
That's shiny!  Thanks @thomasgibson!